src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -191,15 +280,144 @@ static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showF
+        Set<Index> successfulSearchShardIndices,
+        Boolean showFieldName,
+        Boolean showFieldType
+    ) {
         List<String> fieldDataList = new ArrayList<>();
         if (builder instanceof WithFieldName) {
-            fieldDataList.add(((WithFieldName) builder).fieldName());
+            String fieldName = ((WithFieldName) builder).fieldName();
+            if (showFieldName) {
+                fieldDataList.add(fieldName);
+            }
+            if (showFieldType) {
+                String fieldType = getFieldType(fieldName, successfulSearchShardIndices);
+                if (fieldType != null && !fieldType.isEmpty()) {
+                    fieldDataList.add(fieldType);
+                }
+            }
         }
         return " [" + String.join(", ", fieldDataList) + "]";
     }
+
+    String getFieldType(String fieldName, Set<Index> successfulSearchShardIndices) {
+        if (successfulSearchShardIndices == null) {
+            return null;
+        }
+        String fieldType = getFieldTypeFromCache(fieldName, successfulSearchShardIndices);
+        if (fieldType != null) {
+            if (fieldType.equals(NO_FIELD_TYPE_VALUE)) {
+                return null;
+            } else {
+                return fieldType;
+            }
+        }
+
+        Map<String, MappingMetadata> allIndicesMap = getStringMappingMetadataMap(successfulSearchShardIndices);
+        if (allIndicesMap == null) return null;
+
+        return findFieldTypeInMappings(fieldName, successfulSearchShardIndices, allIndicesMap);
+    }
+
+    private String findFieldTypeInMappings(
+        String fieldName,
+        Set<Index> successfulSearchShardIndices,
+        Map<String, MappingMetadata> allIndicesMap
+    ) {
+        for (Index index : successfulSearchShardIndices) {
+            String fieldType = getFieldTypeFromMapping(index, fieldName, allIndicesMap.get(index.getName()));
+            if (fieldType != null) {
+                cacheFieldType(index, fieldName, fieldType);
+                return fieldType;
+            }
+        }
+        return null;
+    }
+
+    String getFieldTypeFromMapping(Index index, String fieldName, MappingMetadata mappingMetadata) {
+        // Get properties from cache or mapping metadata
+        Map<String, ?> propertiesAsMap = propertiesCache.computeIfAbsent(index, k -> {
+            if (mappingMetadata != null) {
+                return (Map<String, ?>) mappingMetadata.getSourceAsMap().get("properties");
+            }
+            return null;
+        });
+
+        // Recursively find the field type from properties
+        if (propertiesAsMap != null) {
+            String fieldType = findFieldTypeInProperties(propertiesAsMap, fieldName.split("\\."), 0);
+
+            // Cache the NO_FIELD_TYPE_VALUE result if no field type is found in this index
+            if (fieldType == null) {
+                cacheFieldType(index, fieldName, NO_FIELD_TYPE_VALUE);
+            }
```

Comment:
Caching should be done IMO. Also, we don't need separate method for this method. Removing unnecessary redirection will make the code more readable